nsqd/protocol_v2.go

```diff
@@ -126,7 +127,28 @@ func (p *protocolV2) SendMessage(client *clientV2, msg *Message) error {
 	p.ctx.nsqd.logf(LOG_DEBUG, "PROTOCOL(V2): writing msg(%s) to client(%s) - %s", msg.ID, client, msg.Body)
 	var buf = &bytes.Buffer{}
 
-	_, err := msg.WriteTo(buf)
+	adjustedBody := msg.Body
+	var err error
+	if msg.snappyCompressed {
+		// message is snappy compressed
+		adjustedBody, err = compress.SnappyDecompress(msg.Body)
+		if err != nil {
+			p.ctx.nsqd.logf(LOG_ERROR, "PROTOCOL(V2): decompress msg(%s) with snappy error", msg.ID)
+			return err
+		}
+		msg.snappyCompressed = false
+	} else if msg.deflateCompressed {
+		// message is deflate compressed
+		adjustedBody, err = compress.DeflateDecompress(msg.Body)
+		if err != nil {
+			p.ctx.nsqd.logf(LOG_ERROR, "PROTOCOL(V2): decompress msg(%s) with deflate error", msg.ID)
+			return err
+		}
+		msg.deflateCompressed = false
+	}
+
+	msg.Body = adjustedBody
+	_, err = msg.WriteTo(buf)
```

Comment:
I'm still confused why the changes here are necessary, the existing compression functionality already delivers compressed messages to consumers?

Reply:
Because the snappy or deflate compression is implemented by wraping the socket, i.e., nsqd will write original message over the wrapped socket and the compressed message will be actually be written to the wire. Reading is the same.

Reply:
Isn't that what we want to happen?

As we had discussed on https://github.com/nsqio/nsq/issues/1148#issuecomment-475809636, I feel like the way to accomplish what you want is for your producers and consumers to compress data.

It probably makes sense to get on the same page about the feature and whether or not it belongs in NSQ core before we discuss code and implementation.

Reply:
Yes. @mreiferson  You're right. I have also consider this pattern about compressing message in producer and decompressing it in consumer. However, On this way we need to do:
1、pass the compression algorithm information to the consumer for it to determine how to decompress it.
2、not all compression algorithm libs are supported by Golang/Python standard lib, for example snappy. In case to decompress the message correctly, consumer should know which compression algorithm the message use or the message can not be consumed correctly and will be requeued or even worse dropped directly once the `max_tries` is reached(I know that we can just print an error log and exit the consumer to indicate that something is wrong, but how about users set the `max_tries` to 1 and the message will be dropped next time it is consumed).

Reply:
How you compress the messages is entirely up to you. In most cases you control both the producer and consumer — is there a need for more than one compression type to be used?

If multiple compression formats do need to be supported, you could also build some identifier into the message structure to indicate to consumers how to decompress.

Reply:
If you have to transition from uncompressed messages to compressed messages, one way to do it is to start a new topic for the compressed messages ... so if you have "events" you might create another topic "events_gzip" and during the transition consume from both. Messages from "events_gzip" are gzipped, in this hypothetical case. But pick whatever is supported by both producer and consumer.

This scheme is nice because nsqd does not have to compress for storage and then decompress for delivery.

Reply:
> If multiple compression formats do need to be supported, you could also build some identifier into the message structure to indicate to consumers how to decompress.

I know this. But not every case we both control producer and consumer nor every producer need add compression byte info into a message. 

> If you have to transition from uncompressed messages to compressed messages, one way to do it is to start a new topic for the compressed messages ... so if you have "events" you might create another topic "events_gzip" and during the transition consume from both. Messages from "events_gzip" are gzipped, in this hypothetical case. But pick whatever is supported by both producer and consumer.

Yes. I agree with this. But what i want to express is to do this transparently to end users(producers and consumers).